sage.py

```diff
@@ -215,259 +171,271 @@ def parse(unparsed_data, alert_labels=[], slim=False):
+
+
+def _group_alerts_per_team(_team_alerts):
+    """Reorganise alerts for each attacker per team"""
+    team_data = dict()
+    for tid, team in enumerate(team_alerts):
+        host_alerts = dict()  # (attacker, victim) -> alerts
+
+        for alert in team:
+            # Alert format: (diff_dt, src_ip, src_port, dst_ip, dst_port, sig, cat, host, ts, mcat)
+            src_ip, dst_ip, signature, ts, mcat = alert[1], alert[3], alert[5], alert[8], alert[9]
+            dst_port = alert[4] if alert[4] is not None else 65000
+            # Simply respect the source,dst format! (Correction: source is always source and dest always dest!)
+
+            # Say 'unknown' if the port cannot be resolved
+            dst_port = 'unknown' if (dst_port not in port_services.keys() or port_services[dst_port] == 'unknown') else port_services[dst_port]['name']
+
+            if (src_ip, dst_ip) not in host_alerts.keys() and (dst_ip, src_ip) not in host_alerts.keys():
+                host_alerts[(src_ip, dst_ip)] = []
+
+            if (src_ip, dst_ip) in host_alerts.keys():
+                host_alerts[(src_ip, dst_ip)].append((dst_ip, mcat, ts, dst_port, signature)) # TODO: remove the redundant host names
+            else:
+                host_alerts[(dst_ip, src_ip)].append((src_ip, mcat, ts, dst_port, signature))
+
+        team_data[tid] = host_alerts.items()
+    return team_data
+
+
+def _get_ups_and_downs(frequencies, slopes):
+    positive = [(0, slopes[0])]  # (index, slope)
+    positive.extend([(i, slopes[i]) for i in range(1, len(slopes)) if (slopes[i - 1] <= 0 and slopes[i] > 0)])
+    negative = [(i + 1, slopes[i + 1]) for i in range(0, len(slopes) - 1) if (slopes[i] < 0 and slopes[i + 1] >= 0)]
+    if slopes[-1] < 0:  # Special case for last ramp down that's not fully gone down
+        negative.append((len(slopes), slopes[-1]))
+    elif slopes[-1] > 0:  # Special case for last ramp up without any ramp down
+        negative.append((len(slopes), slopes[-1]))
+
+    common = set(negative).intersection(positive)
+    negative = [item for item in negative if item not in common]
+    positive = [item for item in positive if item not in common]
+
+    negative = [x for x in negative if (frequencies[x[0]] <= 0 or x[0] == len(frequencies) - 1)]
+    positive = [x for x in positive if (frequencies[x[0]] <= 0 or x[0] == 0)]
+    return positive, negative, common
+
+
+def _plot_episodes(frequencies, episodes, mcat):
+    cap = max(frequencies) + 1
+
+    plt.figure()
+    plt.title(mcat)
+    plt.xlabel('Time ->')
+    plt.ylabel('Frequency')
+    plt.plot(frequencies, 'gray')
+    for ep in episodes:
+        xax_start = [ep[0]] * cap
+        xax_end = [ep[1]] * cap
+        yax = list(range(cap))
+
+        plt.plot(xax_start, yax, 'g', linestyle=(0, (5, 10)))
+        plt.plot(xax_end, yax, 'r', linestyle=(0, (5, 10)))
+
+    plt.show()
+    return
+
 
-## 14
-def legend_without_duplicate_labels(ax, fontsize=10, loc='upper right'):
```

Comment:
Moved this function lower (above the function `_plot_alert_volume_per_episode` where it is called)